### ISSUE TITLE
C++ compile-time array bounds unsuccessfully inferred from template expansions

### ISSUE THREAD

**Issue by mld2443** [llvm/llvm-project#113936]
https://godbolt.org
```cpp
template <int... VALUES>
struct C {
    static constexpr int VALUEARRAY[] = {VALUES...};
};

static_assert(C<0,1,2,3,4>::VALUEARRAY[4] == 4);
```
clang needs the hint of `sizeof...(VALUES)` to build correctly, while GCC and MSVC compile without error.

**Comment by tbaederr** [comment]
The AST node is:
```
`-DeclRefExpr 0x7e0f6b8bcfd0 'const int[]' lvalue Var 0x7e0f6b8bca28 'VALUEARRAY' 'const int[5]' non_odr_use_constant
  `-NestedNameSpecifier TypeSpec 'C<0, 1, 2, 3, 4>':'struct C<0, 1, 2, 3, 4>'
```
But the type of the expression is an `IncompleteArrayType`.

**Comment by shafik** [comment]
CC @cor3ntin

**Comment by frederick-vs-ja** [comment]
Looks overlapping with #79750 to me.

**Comment by mld2443** [comment]
Not exactly, @frederick-vs-ja, #79750 is about the semantics of the types being deduced, and is primarily about clang and GCC being divergent from MSVC.

This issue is specifically about clang being unable to deduce the type at all, being divergent from GCC which can in fact deduce the size of the array correctly without explicit bounds.

**Comment by mld2443** [comment]
I updated the compiler explorer link and description to clarify that the issue is about clang being divergent from both gcc and msvc.

I don't think I have the power to tag, but I would mark this as divergent if I could.
